tls/datadog_checks/tls/tls.py

```diff
@@ -0,0 +1,353 @@
+                self.service_check(
+                    self.SERVICE_CHECK_VALIDATION,
+                    self.CRITICAL,
+                    tags=self._tags,
+                    message='The {} on the certificate does not match the given host'.format(host_type),
+                )
+                return
+            except service_identity.CertificateError as e:  # no cov
+                self.service_check(
+                    self.SERVICE_CHECK_VALIDATION,
+                    self.CRITICAL,
+                    tags=self._tags,
+                    message='The certificate contains invalid/unexpected data: {}'.format(e),
+                )
+                return
+
+        self.service_check(self.SERVICE_CHECK_VALIDATION, self.OK, tags=self._tags)
+
+    def check_age(self, cert):
+        delta = cert.not_valid_after - datetime.utcnow()
+        seconds_left = delta.total_seconds()
+        days_left = seconds_to_days(seconds_left)
+
+        self.gauge('tls.days_left', days_left, tags=self._tags)
+        self.gauge('tls.seconds_left', seconds_left, tags=self._tags)
+
+        if seconds_left <= 0:
+            self.service_check(
+                self.SERVICE_CHECK_EXPIRATION, self.CRITICAL, tags=self._tags, message='Certificate has expired'
+            )
+        elif seconds_left < self._seconds_critical:
+            self.service_check(
+                self.SERVICE_CHECK_EXPIRATION,
+                self.CRITICAL,
+                tags=self._tags,
+                message='Certificate will expire in only {} days'.format(days_left),
+            )
+        elif seconds_left < self._seconds_warning:
+            self.service_check(
+                self.SERVICE_CHECK_EXPIRATION,
+                self.WARNING,
+                tags=self._tags,
+                message='Certificate will expire in {} days'.format(days_left),
+            )
+        else:
+            self.service_check(self.SERVICE_CHECK_EXPIRATION, self.OK, tags=self._tags)
+
+    def create_connection(self):
+        """See: https://github.com/python/cpython/blob/40ee9a3640d702bce127e9877c82a99ce817f0d1/Lib/socket.py#L691"""
+        err = None
+        try:
+            for res in socket.getaddrinfo(self._server, self._port, 0, self._sock_type):
+                af, socktype, proto, canonname, sa = res
+                sock = None
+                try:
+                    sock = socket.socket(af, socktype, proto)
+                    sock.settimeout(self._timeout)
+                    sock.connect(sa)
+                    # Break explicitly a reference cycle
+                    err = None
+                    return sock
+
+                except socket.error as _:
+                    err = _
+                    if sock is not None:
+                        sock.close()
+
+            if err is not None:
+                raise err
+            else:
+                raise socket.error('No valid addresses found, try checking your IPv6 connectivity')
```

Comment:
Is this specifically for IPv6?

Reply:
It's the most common case imo. I couldn't trigger any others